plugins/plugin-hosted/codenvy-machine-hosted/src/test/java/com/codenvy/machine/HostedDockerInstanceTest.java

```diff
@@ -0,0 +1,200 @@
+    private static final String        REGISTRY     = "registry";
+    private static final String        USERNAME     = "username";
+    private static final String        TAG          = "latest";
+    private static final int           CONCURRENCY  = 2;
+    private static final MachineStatus STATUS       = MachineStatus.RUNNING;
+    @Mock
+    private DockerConnector            dockerConnectorMock;
+    @Mock
+    private DockerInstanceStopDetector dockerInstanceStopDetectorMock;
+    @Mock
+    private LineConsumer               outputConsumer;
+    @Mock
+    private DockerNode                 dockerNode;
+
+    private ExecutorService            executor;
+
+    private HostedDockerInstance       dockerInstance;
+
+    private HostedDockerInstance getDockerInstance() {
+        return new HostedDockerInstance(dockerConnectorMock,
+                                        REGISTRY,
+                                        USERNAME,
+                                        mock(DockerMachineFactory.class),
+                                        getMachine(getMachineConfig(true, NAME, TYPE), OWNER, MACHINE_ID, WORKSPACE_ID, STATUS),
+                                        CONTAINER,
+                                        IMAGE,
+                                        dockerNode,
+                                        outputConsumer,
+                                        dockerInstanceStopDetectorMock,
+                                        mock(DockerInstanceProcessesCleaner.class),
+                                        false,
+                                        CONCURRENCY);
+    }
+
+    @BeforeMethod
+    public void setUp() throws IOException {
+        when(dockerNode.getHost()).thenReturn("host1");
+        dockerInstance = getDockerInstance();
+        executor = Executors.newFixedThreadPool(3);
+    }
+
+
+    @Test
+    public void shouldBeAbleToCommitSimultaneously() throws Exception {
+
+        String repo1 = "repo1";
+        String repo2 = "repo2";
+        String repo3 = "repo3";
+
+        WaitingAnswer<Void> waitingAnswer1 = new WaitingAnswer<>();
+        WaitingAnswer<Void> waitingAnswer2 = new WaitingAnswer<>();
+
+        doAnswer(waitingAnswer1).when(dockerConnectorMock).commit(Matchers.argThat(new CommitParamsMatcher(repo1)));
+        doAnswer(waitingAnswer2).when(dockerConnectorMock).commit(Matchers.argThat(new CommitParamsMatcher(repo2)));
+
+        // Starting threads #1 & 2
+        executor.execute(() -> performCommit(repo1, TAG));
+        executor.execute(() -> performCommit(repo2, TAG));
+        waitingAnswer1.waitAnswerCall(1, TimeUnit.SECONDS);
+        waitingAnswer2.waitAnswerCall(1, TimeUnit.SECONDS);
+
+        // when
+        executor.execute(() -> performCommit(repo3, TAG));
+
+        // thread #3 should wait - semaphore is red
+        verify(dockerConnectorMock, never()).commit(Matchers.argThat(new CommitParamsMatcher(repo3)));
```

Comment:
I think that timeout is needed here, otherwise there is no guarantee that 3rd thread was started. Also it would be nice to verify that Instance.commit is called